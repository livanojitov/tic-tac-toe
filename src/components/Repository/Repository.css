.repository-images div{
  width: 23.75%;
  display: inline-block;
  margin-right: 1%;
}

.repository-images div:nth-child(4n+1){
  margin-left: 1%;
}

.repository-images div img{
  max-width: 100%;
} 

.repository select{
  margin: 20px 0;
}

@media screen and (max-width: 400px) {

  .repository-images div{
    width: 48.50%;
  }

  .repository-images div:nth-child(2n+1){
    margin-left: 1%;
  }

}