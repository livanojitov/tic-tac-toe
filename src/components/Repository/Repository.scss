.repository-images {
  div {
    width: 23.75%;
    display: inline-block;
    margin-right: 1%;

    img {
      max-width: 100%;
    }
  }

  div:nth-child(4n + 1) {
    margin-left: 1%;
  }
}

.repository select {
  margin: 20px 0;
}

@media screen and (max-width: 400px) {
  .repository-images {
    div {
      width: 48.5%;
    }

    div:nth-child(2n + 1) {
      margin-left: 1%;
    }
  }
}
