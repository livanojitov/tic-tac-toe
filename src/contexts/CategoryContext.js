import React, { createContext } from 'react';
import useFetchData from '../hooks/useFetchData';

export const CategoryContext = createContext();

const CategoryContextProvider = ({children}) => {
  let {data: categories, error} = useFetchData("categories");

  return (
    <CategoryContext.Provider value={{categories, error}}>
      {children}
    </CategoryContext.Provider>
  );

}
 
export default CategoryContextProvider;