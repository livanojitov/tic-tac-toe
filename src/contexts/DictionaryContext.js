import React, { createContext, useState, useEffect } from 'react';
import axios from 'axios'

export const DictionaryContext = createContext();

const DictionaryContextProvider = (props) => {
  const [DICTIONARY, setDictionary] = useState('');

  useEffect(() => {
    if (DICTIONARY.length === 0){
      axios.get('dictionary.json')
        .then(res => {
          setDictionary(res.data);
        })
    }
  });

  return (
    <DictionaryContext.Provider value={{ DICTIONARY }}>
      {props.children}
    </DictionaryContext.Provider>
  );

}
 
export default DictionaryContextProvider;