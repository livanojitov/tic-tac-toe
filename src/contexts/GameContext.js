import React, { createContext } from 'react';
import useLocalStorage          from '../hooks/useLocalStorage';
export const GameContext = createContext();

const GameContextProvider = ({children}) => {

  const {games, dispatch} = useLocalStorage('tictactoe', []);

  const getGame = (id) => {
    if (games.length > 0){
      return games.filter((game) => game.id === id)[0];
    } else {
      return false;
    }
  }

  return (
    <GameContext.Provider value={{games, dispatch, getGame}}>
      {children}
    </GameContext.Provider>
  );

}
 
export default GameContextProvider;