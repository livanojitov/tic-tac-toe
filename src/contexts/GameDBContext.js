import React, { createContext } from "react";
import useLocalStorage from "../hooks/useLocalStorage";
export const GameDBContext = createContext();

const GameDBContextProvider = ({ children }) => {
  const { games, dispatch } = useLocalStorage("tictactoe", []);

  const getGame = (id) => {
    if (games.length > 0) {
      return games.filter((game) => game.id === id)[0];
    } else {
      return false;
    }
  };

  return (
    <GameDBContext.Provider value={{ games, dispatch, getGame }}>
      {children}
    </GameDBContext.Provider>
  );
};

export default GameDBContextProvider;
