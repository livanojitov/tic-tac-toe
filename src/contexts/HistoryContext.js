import React, { createContext, useState } from 'react';

export const HistoryContext = createContext();

const HistoryContextProvider = (props) => {

  const [history, setHistory] = useState(0);

  const changeHistory = (history) => {
    setHistory(history);
  }

  return (
    <HistoryContext.Provider value={{ history, changeHistory }}>
      {props.children}
    </HistoryContext.Provider>
  );

}
 
export default HistoryContextProvider;