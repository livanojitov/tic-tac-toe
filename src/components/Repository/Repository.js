import React, { useContext, useEffect } from 'react';
import { CategoryContext }   from '../../contexts/CategoryContext';
import Category              from '../Category/Category';
import useCategory           from '../../hooks/useCategory';
import './Repository.css';

const Repository = () => {
  const { categories, error } = useContext(CategoryContext);
  const {category, changeCategoryHandler} = useCategory();
  const url = 'https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images';

  useEffect(() => {
    return () => {
      const chk = document.querySelector('input#menu');
      if (chk){
        chk.checked = false;
      }
    }
  });

  if (error){
    return <div className="error" data-testid="error">{error}</div>
  }

  return categories ? (
    <div className="repository" data-testid="repository">
      <Category onChangeCategory = {changeCategoryHandler}/>
      <div className="repository-images" data-testid="images">
      {Array(~~20).fill(1).map(
          (value, index) => {
            return (
              <div key={index}>
                <img src={`${url}%2F${category.folder}%2F${index}.jpg?alt=media`} alt="user" data-testid="image"/>
              </div>
            )
          })}
      </div>
    </div>
  ) : (
    <div className="loading" data-testid="loading">Loading the Repository page...</div>
  )
}

export default Repository;