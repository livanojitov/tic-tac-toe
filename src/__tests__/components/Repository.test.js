import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { CategoryContext }   from '../../contexts/CategoryContext';
import Repository from '../../components/Repository/Repository';

jest.mock('../../components/Category/Category', () => () => <select />);

describe('Repository component', () => {
  let categories;
  let error;

  beforeEach(() => {
    categories = [{ folder: "habana"    },
                  { folder: "montreal"  },
                  { folder: "vancouver" },
                  { folder: "spain"     },
                  { folder: "germany"   },
                  { folder: "soccer"    },
                  { folder: "fruits"    },
                  { folder: "animals"   },
                  { folder: "puppies"   },
                  { folder: "seinfeld"  }];
    error = "";
  });

  afterEach(cleanup);

  const renderRepository = () => {
    return render(
        <CategoryContext.Provider value={{categories, error}}>
          <Repository/>
        </CategoryContext.Provider>
    );
  }

  test("renders the Repository component correctly", () => {
    const { asFragment } = renderRepository();
    expect(asFragment()).toMatchSnapshot();
  })

  test("categories are not fetched yet", () => {
    categories = "";
    const { queryByTestId } = renderRepository();

    expect(queryByTestId('repository')).toBeNull();
    expect(queryByTestId('loading')).not.toBeNull();
  })

  test("repository class name", () => {
    const { getByTestId } = renderRepository();
    expect(getByTestId('repository')).toHaveClass('repository');
  })
  
  test("repository-images class name", () => {
    const { getByTestId } = renderRepository();
    expect(getByTestId('images')).toHaveClass('repository-images');
  })

  test("error fetching the categories.json file", () => {
    categories = "";
    error = "Error while fetching the categories.json file from the Google Cloud";
    const { getByTestId, queryByTestId } = renderRepository();

    expect(queryByTestId('repository')).toBeNull();
    expect(getByTestId('error')).toHaveTextContent(error);
  })

  test("number of images", () => {
    const { getByTestId, getAllByTestId } = renderRepository();

    expect(getByTestId('images').children.length).toBe(20);
    expect(getAllByTestId('image').length).toBe(20);
  })

  test("images url", () => {
    const { getAllByTestId } = renderRepository();
    const images = getAllByTestId('image');
    const url = 'https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images';

    for (let i=0; i< images.length; i++){
      expect(images[i]).toHaveAttribute('src', expect.stringContaining(url));
    }
    
  })
});
