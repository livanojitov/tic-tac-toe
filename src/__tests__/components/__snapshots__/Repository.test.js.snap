// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository component renders the Repository component correctly 1`] = `
<DocumentFragment>
  <div
    class="repository"
    data-testid="repository"
  >
    <select />
    <div
      class="repository-images"
      data-testid="images"
    >
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F0.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F1.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F2.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F3.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F4.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F5.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F6.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F7.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F8.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F9.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F10.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F11.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F12.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F13.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F14.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F15.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F16.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F17.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F18.jpg?alt=media"
        />
      </div>
      <div>
        <img
          alt="user"
          data-testid="image"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F19.jpg?alt=media"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
